Do not let EventsTimetable schedule past events if catchup=False

avkirilishin opened the pull request:
closes: https://github.com/apache/airflow/issues/33948

Fixed the bug and added some more tests.

avkirilishin committed:
Fix the EventsTimetable schedules past events bug

avkirilishin committed:
Add the empty timetable tests

Taragolis commented on airflow/timetables/events.py:
```diff
@@ -79,22 +82,34 @@ def next_dagrun_info(
-                return None
+        earliest = restriction.earliest
+        if not restriction.catchup:
+            current_time = pendulum.DateTime.utcnow()
```
I would recommend to use `airflow.utils.timezone.utcnow` for this one instead of direct use `pendulum.DateTime.utcnow()`.

We might potentially get rid of `pendulum` in the future, and this might help to replace this easier

avkirilishin committed:
Replace pendulumDateTime by airflow.utils.timezone.

avkirilishin committed:
Add types.

jscheffl reviewed (approved):
Did a regressio with the example included in the bug report and this PR fixed it.

uranusjr committed:
Clean up typing

By coercing things to Pendulum, we can save a bunch of ignores and
explicit annotations.

uranusjr commented on airflow/timetables/events.py:
```diff
@@ -79,22 +83,34 @@ def next_dagrun_info(
+                break
+            next_event = next(dates, None)
+        if next_event is None:
+            return None
```
```suggestion
        if not self.event_dates:
            return None
        for next_event in self.event_dates:
            if earliest and next_event < earliest:
                continue
            elif last_automated_data_interval and next_event <= last_automated_data_interval.end:
                continue
            break
```

Not exactly sure but this seems to be logically equivalent?

avkirilishin committed:
Make minimal changes to the implementation.
